### PR TITLE
Implemented rays refraction

### PR DESCRIPTION
**Pull request by NikitaSukhanov**
### What's new:

- Implemented rays refraction

### Changes: 

- Added `refraction` function, which calculates the refracted ray's direction
- Replaced `while` loop with recursive `trace_ray_rec` function
- Added `refraction` and `transparency` properties for scene objects

### CODE REVIEW

File: raytracing.py [hunk 1]
```diff
@@ -86,8 +86,7 @@ def trace_ray(rayO, rayD):
     toL = normalize(L - M)
     toO = normalize(O - M)
     # Shadow: find if the point is shadowed or not.
-    l = [intersect(M + N * .0001, toL, obj_sh)
-            for k, obj_sh in enumerate(scene) if k != obj_idx]
+    l = [intersect(M + N * .0001, toL, obj_sh) for k, obj_sh in enumerate(scene) if k != obj_idx]
```
Comment (daniilsavchuk) [inline raytracing.py]: what a magic constant? Please define it as constant or as a parameter

**Review by daniilsavchuk — CHANGES_REQUESTED** [review]
Good job, make requested changes and go ahead

File: raytracing.py (hunk 1 above)
Comment (NikitaSukhanov) [reply raytracing.py]: Fixed, it was moved as a constant to the top of the file. Also added it's description.

File: raytracing.py [hunk 2]
```diff
@@ -4,6 +4,11 @@
 w = 400
 h = 300
 
+# Calculating the intersection of the ray, starting from some point of the object's surface, with other objects,
+# we need to offset the ray's start point by normal from the surface.
+# Otherwise the ray may intersect current object in it's starting point.
+surface_offset = .0001
```
Comment (daniilsavchuk) [inline raytracing.py]: Constants according to a PEP8 should be capitalized

**Review by daniilsavchuk — APPROVED** [review]
Fix style and go ahead

File: raytracing.py (hunk 2 above)
Comment (NikitaSukhanov) [reply raytracing.py]: Fixed

**Pull request merged**